src/vs/workbench/browser/parts/panel/panelPart.ts

```diff
@@ -0,0 +1,159 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import {TPromise} from 'vs/base/common/winjs.base';
+import timer = require('vs/base/common/timer');
+import uuid = require('vs/base/common/uuid');
+import strings = require('vs/base/common/strings');
+import {Registry} from 'vs/platform/platform';
+import {IPanel} from 'vs/workbench/common/panel';
+import {EventType as WorkbenchEventType, CompositeEvent} from 'vs/workbench/common/events';
+import {CompositePart} from 'vs/workbench/browser/parts/compositePart';
+import {Panel, PanelRegistry, Extensions as PanelExtensions} from 'vs/workbench/browser/panel';
+import {IPanelService} from 'vs/workbench/services/panel/common/panelService';
+import {IPartService} from 'vs/workbench/services/part/common/partService';
+import {IStorageService, StorageScope} from 'vs/platform/storage/common/storage';
+import {IContextMenuService} from 'vs/platform/contextview/browser/contextView';
+import {IEventService} from 'vs/platform/event/common/event';
+import {IMessageService, Severity} from 'vs/platform/message/common/message';
+import {ITelemetryService} from 'vs/platform/telemetry/common/telemetry';
+import {IKeybindingService} from 'vs/platform/keybinding/common/keybindingService';
+
+export class PanelPart extends CompositePart<Panel> implements IPanelService {
+
+	public static activePanelSettingsKey = 'workbench.panelpart.activepanelid';
+
+	public serviceId = IPanelService;
+
+	private blockOpeningPanel: boolean;
+	private currentPanelOpenToken: string;
+
+	constructor(
+		messageService: IMessageService,
+		storageService: IStorageService,
+		eventService: IEventService,
+		telemetryService: ITelemetryService,
+		contextMenuService: IContextMenuService,
+		partService: IPartService,
+		keybindingService: IKeybindingService,
+		id: string
+	) {
+		super(messageService, storageService, eventService, telemetryService, contextMenuService, partService, keybindingService,
+			(<PanelRegistry>Registry.as(PanelExtensions.Panels)), PanelPart.activePanelSettingsKey, id);
+	}
+
+	public openPanel(id: string, focus?: boolean): TPromise<Panel> {
+		if (this.blockOpeningPanel) {
```

Comment:
@isidorn I would try to push this down to the compositePart because this seems like lots of duplicated code also in the panelPart. You would then also have just one onCompositeOpen event instead of onViewletOpen and onPanelOpen.